### PR TITLE
Improve JITServer AOT cache implementation portability

### PR DESCRIPTION
**Pull request by cjjdespres**
- The `AOTCacheRecord::setSubrecordPointers` method is now protected, instead of private. This works around a bug in clang. Without this, the invocation of `setSubrecordPointers` in the `AOTCacheRecord::readRecord<>` template method fails; clang complains that `setSubrecordPointers` is a private method of `AOTCacheRecord`. This shouldn't matter at all, as a method of `AOTCacheRecord` ought to be able to call the private methods of the `AOTCacheRecord` class itself, but clang issues an erroneous access error anyway.

- Instead of having an `AOTCacheListRecord` class from which the `AOTCacheClassChainRecord` and `AOTCacheWellKnownClassesRecord` classes are derived, those classes now derive from the base `AOTCacheRecord` class and declare their own methods separately. Some code duplication was avoided by adding static template functions for use in those classes' method implementations.

  The previous class inheritance structure was not portable; both clang and gcc support the non-standard flexible array member extension, but only gcc supports deriving from classes with a flexible array member (under certain circumstances).

### CODE REVIEW

**Comment by cjjdespres** [comment]
Attn @mpirvu.

File: runtime/compiler/runtime/JITServerAOTCache.cpp [hunk 1]
```diff
@@ -245,36 +245,35 @@ AOTCacheMethodRecord::subRecordsDo(const std::function<void(const AOTCacheRecord
    f(_definingClassRecord);
    }
 
-template<class D, class R, typename... Args>
-AOTCacheListRecord<D, R, Args...>::AOTCacheListRecord(uintptr_t id, const R *const *records,
-                                                      size_t length, Args... args) :
-   _data(id, length, args...)
+
```
Comment (mpirvu) [inline runtime/compiler/runtime/JITServerAOTCache.cpp:248]: Could you please document these functions?

**Review by mpirvu — APPROVED** [review]
LGTM. It would be nice to add documentation for the new functions.

**Comment by mpirvu** [comment]
Could you please squash the 4 commits into one?

**Comment by mpirvu** [comment]
jenkins test sanity xlinuxjit,zlinuxjit,alinux64jit jdk17

**Pull request merged**